src/psfmachine/tpf.py

```diff
@@ -283,6 +294,130 @@ def plot_tpf(self, tdx):
+        quarter = self.tpf_meta["quarter"][0]
+        # use FFI models
+        if input is None:
+            input = "%s/data/ffi_prf_models_v0.1.0.csv" % (PACKAGEDIR)
+
+        # check if file exists
+        if not os.path.isfile(input):
+            raise FileNotFoundError("No PSF files: ", input)
+
+        try:
+            tab = pd.read_csv(input, index_col=0, header=[0, 1])
+            self.n_r_knots = int(tab.loc[channel, (str(quarter), "n_r_knots")])
+            self.n_phi_knots = int(tab.loc[channel, (str(quarter), "n_phi_knots")])
+            # *4 to convert from pixel to arcseconds
+            self.rmin = int(tab.loc[channel, (str(quarter), "rmin")] * 4)
+            self.rmax = int(tab.loc[channel, (str(quarter), "rmax")] * 4)
+            self.cut_r = int(6 * 4)
+            self.psf_w = tab.loc[channel, str(quarter)].iloc[4:].values
+            # self.psf_w_err = tab.loc[channel, str(quarter)].iloc[4:].values
+            del tab
+        # if channel or quarter data is not in file, then a KeyError is raised
+        except KeyError:
+            raise IOError(
+                "Quarter %i and channel %i has no PRF model data" % (quarter, channel)
+            )
+
+        # create source mask and uncontaminated pixel mask
+        self._get_source_mask()
+        self._get_uncontaminated_pixel_mask()
+
+        # create mean model, but PRF shapes from FFI are in pixels! and TPFMachine
+        # work in arcseconds
+        self._get_mean_model()
+
+        # Re-estimate source flux
+        # -----
+        prior_mu = self.source_flux_estimates
+        prior_sigma = (
+            np.ones(self.mean_model.shape[0]) * 10 * self.source_flux_estimates
+        )
+
+        f, fe = (self.flux).mean(axis=0), ((self.flux_err ** 2).sum(axis=0) ** 0.5) / (
+            self.nt
+        )
+
+        X = self.mean_model.copy()
+        X = X.T
+
+        sigma_w_inv = X.T.dot(X.multiply(1 / fe[:, None] ** 2)).toarray()
+        sigma_w_inv += np.diag(1 / (prior_sigma ** 2))
+        B = X.T.dot((f / fe ** 2))
+        B += prior_mu / (prior_sigma ** 2)
+        ws = np.linalg.solve(sigma_w_inv, B)
+        werrs = np.linalg.inv(sigma_w_inv).diagonal() ** 0.5
+
+        # -----
+
+        # Rebuild source mask
+        ok = np.abs(ws - self.source_flux_estimates) / werrs > 3
+        ok &= ((ws / self.source_flux_estimates) < 10) & (
+            (self.source_flux_estimates / ws) < 10
+        )
+        ok &= ws > 10
+        ok &= werrs > 0
+
+        self.source_flux_estimates[ok] = ws[ok]
+
+        self.source_mask = (
+            self.mean_model.multiply(
+                self.mean_model.T.dot(self.source_flux_estimates)
+            ).tocsr()
+            > 1
+        )
+
+        # Recreate uncontaminated mask
+        self._get_uncontaminated_pixel_mask()
+
+        # Recreate mean model!
+        self._get_mean_model()
+
+        if plot:
+            return self.plot_shape_model()
+        return
```

Comment:
From line 330-386: I think this little loop (or something similar) is also undertaken in `build_shape_model`. I think it's removing "background" pixels. We should pull this out as it's own helper method, so that we're not repeating ourselves.

Reply:
implemented in commit:
85ccd6461f95714826a9f64472cc0433942780d6